When available use AnnotationRegistry::registerUniqueLoader

jrjohnson opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 3.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| References tickets | #24596
| License       | MIT

Take advantage of AnnotationRegistry::registerUniqueLoader which will only add 'class_exists' as an autoloader if it has not already been added. This helps alleviate a performance issue when the
same loader is added many times in tests.

jrjohnson committed:
Take advantage of AnnotationRegistry::registerUniqueLoader

This method will only add 'class_exists' as an autoloader if it has not
already been added. This helps alleviate a performance issue when the
same loader is added many times in tests.

nicolas-grekas commented on src/Symfony/Bridge/PhpUnit/Legacy/SymfonyTestsListenerTrait.php:
```diff
@@ -120,7 +120,11 @@ public function startTestSuite($suite)
 
             if (!class_exists('Doctrine\Common\Annotations\AnnotationRegistry', false) && class_exists('Doctrine\Common\Annotations\AnnotationRegistry')) {
-                AnnotationRegistry::registerLoader('class_exists');
+                if (method_exists(AnnotationRegistry::class, 'registerUniqueLoader')) {
```
The bridge is still PHP 5.3 compatible so this should be replaced by a regular string instead of:: class

nicolas-grekas replied:
But: does this actually do something? The first class_exists check just above already ensures unicity, isn't it?

jrjohnson replied:
The class now has two methods. In v1.6+ `registerUniqueLoader` exists and should be used, but it requires PHP 7.1+.

nicolas-grekas reviewed (dismissed):
Once comment is addressed.

nicolas-grekas commented:
There is another call to registerLoader in an XML file in FrameworkBundle, should be replaced also I believe when possible.

jrjohnson commented:
I split into two PRs since this one could maintain BC and the change in the XML could not (as far as I can tell).

stof commented:
Changing the registration in the XML file means you should alter the Definition object from the DI extension instead of registering it from the XML, to be able to choose the one being registered

stof commented on src/Symfony/Bridge/PhpUnit/Legacy/SymfonyTestsListenerTrait.php:
```diff
@@ -269,7 +273,7 @@ public function endTest($test, $time)
             foreach ($deprecations ? unserialize($deprecations) : array() as $deprecation) {
                 if ($deprecation[0]) {
-                    trigger_error(serialize(array('deprecation' => $deprecation[1], 'class' => $className, 'method' => $test->getName(false))), E_USER_DEPRECATED);
+                    @trigger_error(serialize(array('deprecation' => $deprecation[1], 'class' => $className, 'method' => $test->getName(false))), E_USER_DEPRECATED);
```
this looks wrong to me, as it means the `if` and `else` blocks are now exactly the same.

nicolas-grekas replied:
that was suggested by fabbot, but should be reverted indeed, that's a false-positive

jrjohnson commented:
Thanks @stof and @nicolas-grekas. I reverted the false positive code style fix and moved the invocation on `AnnotationRegistry` into the configuration PHP. I'm not sure that is the right place for it so please let me know if I went the wrong direction.

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/annotations.xml:
```diff
@@ -7,20 +7,6 @@
-                <argument type="service">
-                    <!-- dummy arg to register class_exists as annotation loader only when required -->
-                    <service class="Doctrine\Common\Annotations\AnnotationRegistry">
-                        <call method="registerLoader">
```
you should keep this definition, and just replace registerLoader by registerUniqueLoader here

THEN
in FrameworkExtension.php you need to fetch the "annotations.reader" definition using `$container->getDefinition('annotations.reader')`, and tweak it as needed to replace the call by "registerLoader" when the previous method doesn't exist.

nicolas-grekas replied:
Something like:
`$container->getDefinition('annotations.reader')->getArgument(1)->setMethodCalls(array('registerLoader', array('class_exists')));`

jrjohnson replied:
Ah ok, thanks. I would never have thought of that. I will make that change now.

jrjohnson replied:
Sorry @nicolas-grekas this is beyond my skill and understanding. `$container->getDefinition('annotations.reader')
    ->getMethodCalls()[0][1][1]` returns a `Symfony\Component\DependencyInjection\Reference` which doesn't have a `setMethodCalls` method. That method does exist on the `Definition` but I'm not sure how to duplicate the tiered structure in the XML or the call to a service.

nicolas-grekas replied:
Instead of an inlined service, you may move the service on L15 as a named one. That may help you, and future readers.

jrjohnson replied:
Looked back at the original code and doing this as a configurator was the suggested approach.  Since I couldn't figure out how to do this another way I did it with a configurator. This has the added benefit of calling at static methods statically and not relying oh PHPs poor error control around calling them as instance methods.
 I wasn't sure where to put the class though. I didn't want to create a special namespace just for it, so I shoehorned it into CacheWarmer which is probably doesn't belong.

nicolas-grekas replied:
That's reasonable, but actually we made it this way on purpose: to not add yet another class just for this.
I just pushed the change that makes this works as suggested :)

jrjohnson commented:
Thanks for the assist @nicolas-grekas! A rebase seems to have cleared up the test issues so I believe this is ready.

fabpot commented:
Thank you @jrjohnson.
